module/zfs/zio.c

```diff
@@ -1005,6 +1029,174 @@ zio_ioctl(zio_t *pio, spa_t *spa, vdev_t *vd, int cmd,
 	return (zio);
 }
 
+/*
+ * Performs the same function as zio_trim_tree, but takes a dkioc_free_list_t
+ * instead of a range tree of extents. The `dfl' argument is stored in the
+ * zio and shouldn't be altered by the caller after calling zio_trim_dfl.
+ * If `dfl_free_on_destroy' is true, the zio will destroy and free the list
+ * using dfl_free after the zio is done executing.
+ */
+zio_t *
+zio_trim_dfl(zio_t *pio, spa_t *spa, vdev_t *vd, dkioc_free_list_t *dfl,
+    boolean_t dfl_free_on_destroy, boolean_t auto_trim,
+    zio_done_func_t *done, void *private)
+{
+	zio_t *zio;
+	int c;
+
+	ASSERT(dfl->dfl_num_exts != 0);
+
+	if (vd->vdev_ops->vdev_op_leaf) {
+		/*
+		 * A trim zio is a special ioctl zio that can enter the vdev
+		 * queue. We don't want to be sorted in the queue by offset,
+		 * but sometimes the queue requires that, so we fake an
+		 * offset value. We simply use the offset of the first extent
+		 * and the minimum allocation unit on the vdev to keep the
+		 * queue's algorithms working more-or-less as they should.
+		 */
+		uint64_t off = dfl->dfl_exts[0].dfle_start;
+
+		zio = zio_create(pio, spa, 0, NULL, NULL, 1 << vd->vdev_ashift,
+		    1 << vd->vdev_ashift, done, private, ZIO_TYPE_IOCTL,
```

Comment:
So the issue here is that ashift gets stored in the label for the top-level vdev.  That can differ from device reported ashift which gets set during the export/import.  What I think we should be doing here is passing `vd->vdev_top->vdev_ashift` to ensure it's aligned correctly for all leaf vdevs.  Even though the zio itself contains no ABD since it's a TRIM.  @skiselkov what do you think of this fix?

```diff
diff --git a/module/zfs/zio.c b/module/zfs/zio.c
index 228cb33..e220df5 100644
--- a/module/zfs/zio.c
+++ b/module/zfs/zio.c
@@ -1064,9 +1064,10 @@ zio_trim_dfl(zio_t *pio, spa_t *spa, vdev_t *vd, dkioc_fr
                 * queue's algorithms working more-or-less as they should.
                 */
                uint64_t off = dfl->dfl_exts[0].dfle_start;
+               uint64_t size = 1 << vd->vdev_top->vdev_ashift;
 
-               zio = zio_create(pio, spa, 0, NULL, NULL, 1 << vd->vdev_ashift,
-                   1 << vd->vdev_ashift, done, private, ZIO_TYPE_IOCTL,
+               zio = zio_create(pio, spa, 0, NULL, NULL,
+                   size, size, done, private, ZIO_TYPE_IOCTL,
                    auto_trim ? ZIO_PRIORITY_AUTO_TRIM : ZIO_PRIORITY_MAN_TRIM,
                    ZIO_FLAG_CANFAIL | ZIO_FLAG_DONT_RETRY |
                    ZIO_FLAG_DONT_PROPAGATE | ZIO_FLAG_DONT_AGGREGATE, vd, off,
```